Refactor/#101 auth request

soralee2821 opened the pull request:
## Issues
- Issue number #101 

## Tasks Done 
- [x] Authorization Request 관련 리팩터링
  - [x] `instance` 파일에서 불필요한 콜백 함수 정의 제거
  - [x] `useAuthApi` 파일의 함수들의 에러 처리 로직을 `excuteAuthRequestWithErrorHandling` 함수로 분리
- [x] Modal 관련 리팩터링 
  - [x]  Modal 파일들을 `components/modals` 폴더로 이동
  - [x] 모달 타입을 나타내는 전역 상태 **modalTypeAtom** 선언
    - [x] 기존의 모달들이 **showModal** 대신 **modalTypeAtom**을 사용하도록 수정
    - [x] `Root` 페이지에서 **modalTypeAtom**에 맞는 모달을 조건부 렌더링함
- [x] HTTP 상태 코드 **401** 에러 응답이 왔을 때 에러 처리 
  - [x] `excuteAuthRequestWithErrorHandling` 함수에서 **authRequest** 함수를 호출했을 때 **401** 에러 응답이 왔을 때 에러 처리
    - [x] code **A02**일 때, `reissueAccessToken` 함수를 호출하여 재발급 받은 **accessToken**을 사용하여  **authRequest** 함수 재호출
  - [x] `reissueAccessToken` 함수 호출에 대해 HTTP 상태 코드 **401** 에러 응답이 왔을 경우 에러 처리
    - [x] code **A02**일 때, "로그인 기한 만료되었습니다." 모달창을 띄움
      - [x] "확인" 버튼 클릭 시, 로그인 페이지로 이동 
    - [x] code **A03**일 때, "계정이 도용된 것으로 의심됩니다." 모달창을 띄움
      - [x] "확인" 버튼 클릭 시, 로그아웃 로직 실행 

<br>

## Description
### Authorization Request 관련 리팩터링
- `axios/instance` 파일에서 불필요한 **intercepter**의 콜 함수들을 제거했습니다.
- **Authorization Request**의 에러 처리 로직이 같으므로, 에러 처리 로직을 `excuteAuthRequestWithErrorHandling` 함수로 추출했습니다.
  - `excuteAuthRequestWithErrorHandling` 함수는 실제 호출하는 API 함수를 **authRequest** 매개변수로 받습니다.
  - **try~catch** 문 내부에서 **authRequest** 함수를 호출하고 에러가 발생할 경우 **catch**문에서 에러를 해결합니다.
    - **Access Token**이 만료되었을 경우, **A02** 코드의  HTTP 상태 코드 **401** 에러가 발생합니다.
    - `initializeAccessToken` 함수를 통해 재발급 받은 **Access Token**을 사용하여  **authRequest** 함수를 다시 호출합니다.

<br>

### Modal 관련 리팩터링 및 LoginExpirationModal, AccountHijacking 모달 구현
- 모달 컴포넌트들을 `components/modals` 폴더 내로 이동했습니다.
  - 모달 컴포넌트에서 사용하는 상수들을 해당 컴폰넌트 파일로 이동했습니다.
- 모달의 타입을 나타내는 전역 상태 **modalTypeAtom**을 선언했습니다.
  - 각 모달 컴포넌트에서 부모 컴포넌트의 **showModal** 대신에 **modalTypeAtom**을 사용하도록 수정했습니다.
  - `Root` 페이지에서 **modalTypeAtom**에 따라 해당 모달 컴포넌트를 조건부 렌더링했습니다.
- **Refresh Token**이 만료되었을 때, 사용자에게 로그인이 만료되었음을 명시하고 로그인 페이지로 이동하도록 유도하는`LoginExpirationModal`을 구현했습니다.
- **Refresh Token**이 도용되었을 때, 사용자에게 계정 도용이 의심됨을 명시하고 로그아웃을 유도하는`AccountHijackingModal`을 구현했습니다.

<br>

### 토큰 재발급 API 함수 HTTP 상태 코드 **401** 에러 응답이 왔을 때 에러 처리 
- `reissueAccessToken` 함수 호출에 대해 **401** 에러 응답이 왔을 경우 에러 처리를 구현했습니다.
- **401**에러 응답과 관련된 모달을 띄우기 위해 **setModalType**을 `reissueAccessToken` 함수의 매개변수로 설정했습니다.
  - `reissueAccessToken` 함수를 호출하는 `initializeAccessToken` 함수의 인수로 **setModalType**을 전달했습니다.
  - `initializeAccessToken` 함수를 호출하는 `excuteAuthRequestWithErrorHandling` 함수의 인수로 **setModalType**을 전달했습니다.
  - `excuteAuthRequestWithErrorHandling` 함수를 호출하는 **User API** 함수인 `deleteAccountWithAuthenticationCode`, `sendAuthenticationCode`, `validateNickname` 함수의 인수로 **setModalType**을 전달했습니다.
  - `excuteAuthRequestWithErrorHandling` 함수를 호출하는 **User** 관련 **Tanstack-query** 함수인 `useGetUserData`, `usePostNickname` 함수의 인수로 **setModalType**을 전달했습니다.
  - `excuteAuthRequestWithErrorHandling` 함수를 호출하는 **Workout** 관련 **Tanstack-query** 함수인 `useGetWorkoutData`, `usePostWorkoutData`  함수의 인수로 **setModalType**을 전달했습니다.
- **setModalType** 사용하여, 에러 응답과 관련된 모달을 렌더링 하도록 구현했습니다.
  - **A02** 코드의 **401** 에러 응답이 왔을 경우, **setModalType**에 *loginExpiration* 값을 전달하여 `LoginExpirationModal`을 렌더링 했습니다.
  - **A03** 코드의 **401** 에러 응답이 왔을 경우, **setModalType**에 *accountHijacking* 값을 전달하여 `LoginExpirationModal`을 렌더링 했습니다.

<br>

### 기타 리팩터링
- input value의 의미를 드러내기 위해, `EditMyPage`의 `INPUT_LENTH`를 `NICKNAME_LENGTH`로 수정했습니다.
- **Tanstack-query** 관련 User 함수명에서 HTTP method가 잘 드러나도록 `useUserData`를 `useGetUserData`로, `useMutationNickname`을 `usePostNickname`으로 변경했습니다.
- `Root` 페이지에서 `initializeAccessToken` 함수의 인수로 **setModalType**을 전달했습니다.
- 로그아웃했을 때 **Access Token**이 ''이 되므로, `Root` 페이지의 **useEffect** 조건문에서의 **accessToken**에 optional operator(**?**)을 추가했습니다.
- `reissueAccessToken` 함수의 에러 처리 로직에서 **A02**, **A03** 문자열 대신에 `RESPONSE_CODE` 상수를 선언하고 사용해서 의도를 드러내었습니다.

## Visuals

#### LoginExpirationModal
![loginExpirationModal](https://github.com/Pullanner/pullanner-web/assets/70058081/f2f6e1ab-77a1-4954-b63e-96db5aab2630)

#### AccountHijackingModal
![accountHijackingModal](https://github.com/Pullanner/pullanner-web/assets/70058081/ad7a4508-6cb8-4c74-a5fc-dfb7795d7b7f)

soralee2821 committed:
[#101] Refactor: remove unnecessary interceptors

soralee2821 committed:
[#101] Feat: add func to excute authRequest and handle error

soralee2821 committed:
[#101] Refactor: remove error handling logic from getAuthRequest

soralee2821 committed:
[#101] Refactor: change getAuthRequest to excuteAuthRequestWithErrorHandling

- validateNickname 함수의 getAuthRequest 함수를 excuteAuthRequestWithErrorHandling로 변경함
- useUserData 함수의 getAuthRequest 함수를excuteAuthRequestWithErrorHandling로 변경함
- useGetWorkoutData 함수의 getAuthRequest 함수를excuteAuthRequestWithErrorHandling로 변경함

soralee2821 committed:
[#101] Refactor: remove error handling logic from postAuthRequest

soralee2821 committed:
[#101] Refactor: change postAuthRequest to excuteAuthRequestWithErrorHandling

- sendAuthenticationCode 함수의 postAuthRequest 함수를 excuteAuthRequestWithErrorHandling로 변경함
- useMutationNickname 함수의 postAuthRequest 함수를excuteAuthRequestWithErrorHandling로 변경함
- usePostWorkoutData 함수의 postAuthRequest 함수를excuteAuthRequestWithErrorHandling로 변경함

soralee2821 committed:
[#101] Refactor: remove error handling logic from deleteAuthRequest

soralee2821 committed:
[#101] Refactor: change deleteAuthRequest to excuteAuthRequestWithErrorHandling

- deleteAccountWithAuthorizationCode 함수의 deleteAuthRequest 함수를 excuteAuthRequestWithErrorHandling로 변경함

soralee2821 committed:
[#101] Rename: move Modal folder to modals folders

soralee2821 committed:
[#101] Rename: move existing Modal files to modals folders

soralee2821 committed:
[#101] Rename: move text content of modal into the modal file

soralee2821 committed:
[#101] Rename: change constant name from INPUT_LENGTH to NICKNAME_LENGTH

soralee2821 committed:
[#101] Feat: declare modalTypeAtom

- 모달을 modalType에 따라 다르게 렌더링하기 위해 modalTypeAtom을 선언함
- modalTypeAtom의 setter 함수를 props로 넘겨주기 위해 SetModalType 타입을 선언함

soralee2821 committed:
[#101] Refactor: apply modalTypeAtom to DeleteAccountModal

- EditMyPage의 showModal 상태 대신에 modalTypeAtom을 사용하여 모달 상태를 관리함
- Root 페이지에서 modalType이 'deleteAccount'일 때 해당 모달을 조건부 렌더링함

soralee2821 committed:
[#101] Refactor: apply modalTypeAtom to DeleteAccountSuccessModal

- AuthenticationCodeInput의 showModal 상태 대신에 modalTypeAtom을 사용하여 모달 상태를 관리함
- Root 페이지에서 modalType이 'deleteAccountSuccess'일 때 해당 모달을 조건부 렌더링함

soralee2821 committed:
[#101] Refactor: apply modalTypeAtom to LogoutModal

- MyPage의 showModal 상태 대신에 modalTypeAtom을 사용하여 모달 상태를 관리함
- Root 페이지에서 modalType이 'logout'일 때 해당 모달을 조건부 렌더링함

soralee2821 committed:
[#101] Refactor: modify excuteAuthRequestWithErrorHandling by using initializeAccessToken

- initializeAccessToken 함수가 새로 발급받은 AccessToken을 반환하도록 수정함
- excuteAuthRequestWithErrorHandling 함수가 initializeAccessToken 함수를 사용하도록 수정함

soralee2821 committed:
[#101] Refactor: apply changed apiResponse to error handling logics

- excuteAuthRequestWithErrorHandling 함수에서 AccessToken 만료시 'A02' 코드의 401 에러로 변경함
- ReissueAccessToken 함수에서 RefreshToken 만료시  'A02' 코드의 401 에러로 변경함
- ReissueAccessToken 함수에서 RefreshToken 도용 의심시  'A03' 코드의 401 에러로 변경함

soralee2821 committed:
[#101] Feat: add LoginExpirationModal to modals folder

- modalTypeAtom에 'loginExpiration' 값 추가함
- Root 페이지에서 modalType이 'loginExpiration'일 때 해당 모달을 조건부 렌더링함

soralee2821 committed:
[#101] Feat: add AccountHijackingModal to modals folder

- modalTypeAtom에 'accountHijacking' 값 추가함
- Root 페이지에서 modalType이 'accountHijacking'일 때 해당 모달을 조건부 렌더링함

soralee2821 committed:
[#101] Refactor: add setModalType param to utill funcs

- reissueAccessToken 함수에 setModalType 매개변수를 추가하고, setModalType을 사용하여 각 경우에 대한 modalType을 설정함
- initiaizeAccessToken 함수에 setModalType 매개변수를 추가함
- excuteAuthRequestWithErrorHandling 함수에 setModalType 매개변수를 추가함

soralee2821 committed:
[#101] Refactor: add setModalType param to user api funcs

- deleteAccountWithAuthenticationCode 함수에 setModalType 매개변수를 추가하고, 이 함수를 호출할 때 setModalType을 인수로 넣어줌
- sendAuthenticationCode 함수에 setModalType 매개변수를 추가하고, 이 함수를 호출할 때 setModalType을 인수로 넣어줌
- validateNickname 함수에 setModalType 매개변수를 추가하고, 이 함수를 호출할 때 setModalType을 인수로 넣어줌

soralee2821 committed:
[#101] Refactor: add setModalType param to user react-query funcs

- useUserData 함수에 setModalType 매개변수를 추가하고, 이 함수를 호출할 때 setModalType을 인수로 넣어줌
- useMutateNickname  함수에 setModalType 매개변수를 추가하고, 이 함수를 호출할 때 setModalType을 인수로 넣어줌

soralee2821 committed:
[#101] Rename: modify names of user react-query function to represent HTTP Method

- useUserData 함수명을 useGetUserData로 변경함
- useMutateNickname 함수명을 usePostNickname으로 변경함

soralee2821 committed:
[#101] Refactor: add setModalType param to workout react-query funcs

- useGetWorkoutData 함수에 setModalType 매개변수를 추가함
- usePostWorkoutData  함수에 setModalType 매개변수를 추가하고, 이 함수를 호출할 때 setModalType을 인수로 넣어줌

soralee2821 committed:
[#101] Refactor: add setModalType argument to initializeAccessToken

soralee2821 committed:
[#101] Refactor: add optional operator to accessToken in useEffect

soralee2821 committed:
[#101] Refactor: declare RESPONSE_CODE constant to represent intention

- `reissueAccessToken` 함수의 에러 처리 로직의 조건문에서 'A02', 'A03' 문자열을 RESPONSE_CODE.expiredRefreshToken와 RESPONSE_CODE.hijackedRefreshToken로 변경하여 의도를 드러냄

jaypedia commented on src/lib/axios/executeAuthRequestWithErrorHandling.ts:
```diff
@@ -0,0 +1,41 @@
+  setModalType: SetModalType;
+};
+
+export const excuteAuthRequestWithErrorHandling = async ({
```
앗 함수 명에 오타가.. execute를 의도하신거겠죠? 
함수명이 너무 길기도 해서 `handleAuthRequest` 이런 이름은 어떨까요? 조금 축약해서 표현해도 좋을 것 같아요!

soralee2821 replied:
네, 오타입니다ㅠㅠ `handleAuthRequest`로 바꿔도 좋습니다. 
에러 핸들링도 이름에 포함해야 되지 않을까 싶어서 포함했는데 그러다보니까 이름이 길어졌네요.

jaypedia commented on src/pages/EditMyPage/index.tsx:
```diff
@@ -7,22 +7,22 @@ import { DimmedButton } from '@/components/buttons/DimmedButton';
   const [accessToken, setAccessToken] = useAtom(accessTokenAtom);
-  const { mutate } = useMutateNickname(accessToken, setAccessToken);
+  const setModalType = useSetAtom(modalTypeAtom);
+  const { mutate } = usePostNickname(accessToken, setAccessToken, setModalType);
```
사소한거긴 한데 mutate를 그대로 사용하기보다 의미있는 이름을 붙여서 사용하면 좋을거같아요! 이것도 컨벤션으로 논의하면 좋은데 여기서 이야기하자면, 저같은 경우는 `{mutate: postNickName}` 이런식으로 이름을 붙여요

soralee2821 replied:
저는 `mutate` 그대로 사용하는 편이었는데, 의미있는 이름을 붙여서 사용하는 것도 괜찮아요.
그럼 `mutate`를 `postNickname`으로 이름을 변경할께요.

jaypedia commented on src/pages/Root/index.tsx:
```diff
@@ -32,6 +39,11 @@ export const Root = () => {
+      {modalType === 'deleteAccountSuccess' && <DeleteAccountSuccessModal />}
+      {modalType === 'logout' && <LogoutModal />}
+      {modalType === 'loginExpiration' && <LoginExpirationModal />}
+      {modalType === 'accountHijacking' && <AccountHijackingModal />}
```
모달의 타입이 많아질수록 Root 컴포넌트에 계속 이 로직을 추가하는 방법보다는 좀 더 확장성 있는 방식을 고민해보면 좋을거같아요! 저도 열심히 고민해볼게요!

soralee2821 replied:
저도 확장성 있는 방법이 있는지 더 고민해보겠습니다.

jaypedia commented on src/pages/SelectWorkout/index.tsx:
```diff
@@ -21,7 +22,8 @@ export const SelectWorkout = () => {
   const [accessToken, setAccessToken] = useAtom(accessTokenAtom);
-  const { mutate } = usePostWorkoutData(accessToken, setAccessToken);
+  const setModalType = useSetAtom(modalTypeAtom);
+  const { mutate } = usePostWorkoutData(accessToken, setAccessToken, setModalType);
```
이부분도 전 코멘트에 남긴 내용과 동일합니다!

soralee2821 replied:
`mutate`를 `postWorkoutData`로 수정할께요!

jaypedia commented on src/pages/SetNickname/index.tsx:
```diff
@@ -24,7 +25,8 @@ const DESCRIPTION_TEXT = [
   const [accessToken, setAccessToken] = useAtom(accessTokenAtom);
-  const { mutate } = useMutateNickname(accessToken, setAccessToken);
+  const setModalType = useSetAtom(modalTypeAtom);
+  const { mutate } = usePostNickname(accessToken, setAccessToken, setModalType);
```
이부분도요!

soralee2821 replied:
네, 여기도 동일하게 `mutate`를 `postNickname`으로 이름을 변경할께요.

jaypedia commented on src/lib/axios/executeAuthRequestWithErrorHandling.ts:
```diff
@@ -0,0 +1,41 @@
+      const { status, data } = error.response;
+      if (status === 401 && data.code === 'A02') {
+        const newAccessToken = await initializeAccessToken(setAccessToken, setModalType);
+        const retriedResponse = await authRequest(newAccessToken);
```
retry는 우선 1회만 진행하는 것으로 작성하신거네요!

soralee2821 replied:
네, **Access Token**이 만료되었을 때의 에러 처리여서 **Access Token**을 재발급 받은 후 API 요청을 한 번만 호출해도충분할거라고 생각했습니다.
그리고 `executeAuthRequestWithErrorHandling` 함수 자체를 재귀적으로 호출하지 않은 이유는 그렇게 할 경우 무한히 호출될 위험이 있어서였습니다.

jaypedia reviewed (changes requested):
에러 처리 로직, 전역 모달 등 고민해야 하는 부분이 많은 PR이었네요! 
전역 모달은 저도 고민해보고 같이 논의한 후 최종적으로 적용하면 좋을 것 같습니다~
고생하셨습니다!

soralee2821 committed:
[#101] Rename: change excuteAuthRequestWithErrorHandling to handleAuthRequest

soralee2821 committed:
[#101] Rename: change mutate name to meaning name

- `EditMyPage`와 `SetNickname`에서 mutate를 postNickname으로 변경함
- `SelectWorkout`에서 mutate를 postWorkoutData로 변경함
